### DISCUSSION TITLE
OpenMCT integration

### DISCUSSION THREAD

**Discussion by shalevm** [nasa/fprime#1635] [Q&A] [answered]
Hi all,
I'd like to integrate f-pime with a professional grade ground segment. OpenMCT seems like the right choice, but a few questions arise:
1. Am I correct in my understanding that OpenMCT is only meant to display telemetry, and does not support uplinking commands and/or files? What is the solution in this case for a command/uplink channel? running OpenMCT and GDS simultaneously? 

2. I found [this](https://github.com/fprime-community/fprime-openmct-integration) repo that's supposed to bridge the gap of integrating f-prime and OpenMCT, but it seems extremely out of date (last update from over 3 years ago...), to the point that it's unusable. For instance, in order to produce the dictionary in JSON format for the given deployment XML, it assumes the existence of an autocoder named `JSONDictonaryGen.py` (in the fprime repo), **which does not exist**. Moreover, it seems to also assume a different format for the deployment XML (assumes the XML holds deployment names). 
Does anyone have an up-to-date solution for this integration?
If not - can someone specify a reasonably detailed roadmap of the work needed?

3. While reading the `README.md` file under `fprime/Autocoders/Python` i noticed it describes several autocoders that do not currently exist in the fprime repo, e.g.: 
`JSONDictionaryGen.py` (mentioned above)
`cosmosgen.py`
`gds_dictgen.py `
Is this a matter of the `README.md` "jumping the gun", meaning these will be implemented in the future? or is it not up to date, meaning these have already been deleted from the repo?

Thanks.

**Comment by Joshua-Anderson** [comment, marked as answer]
> Am I correct in my understanding that OpenMCT is only meant to display telemetry, and does not support uplinking commands and/or files?

That's my understanding as well. OpenMCT focuses on the telemetry and visualization side and expects you to use a different tool for commanding.

The fprime openmct integration is quite old and so a lot of the autocoder side of things has changed, though the integration could be a good starting place.

> If not - can someone specify a reasonably detailed roadmap of the work needed?

A lot of this depends on how you intend to use your ground system. The simple solution like the integration demo you linked to wouldn't be too much work. You could use the existing fprime-gds to decode and deserialize telemetry, then either push or poll the deserialized telemetry packets from the GDS into openmct with a simple openmct plugin.

A "flight ready" solution would probably involve coming up with some solution to pipe and store telemetry in a database, the writing a plugin for openmct to query the database, but that's likely significantly more effort

**Reply by shalevm** [reply]
Thanks Joshua!
about the "flight ready" solution - doesn't OpenMCT come with its own database? or did I misunderstand your intent?

**Reply by Joshua-Anderson** [reply]
OpenMCT doesn't come with it's own database - the intension is that the user will write an adapter to connect OpenMCT to their own database. There are some database plugins to allow you to save user created openmct layouts and such to an database.

**Reply by shalevm** [reply]
awesome, thanks!

**Comment by ThibFrgsGmz** [comment]
@shalevm If you manage to set up OpenMCT with FPrime, I would be grateful for a brief summary of the procedure or a tutorial!

I had started to imagine having:
 - `fprime-gds` as a control screen,
 - an `InfluxDb` time series database to record the telemetry, and
 - `OpenMCT` to display the data in real time.

**Reply by shalevm** [reply]
@ThibFrgsGmz  Will do!
(sorry for the late response, just got back from vacation :)  )
Shalev.
